File: leetcode/0297-serialize-and-deserialize-binary-tree.java
import java.util.*;

public class Codec {

    // 직렬화
    public String serialize(TreeNode root) {
        if (root == null) {
            return "# #";
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        List<String> result = new ArrayList<>();
        result.add("#");

        // 트리 BFS 직렬화
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();

            if (node != null) {
                queue.offer(node.left);
                queue.offer(node.right);

                result.add(String.valueOf(node.val));
            } else {
                result.add("#");
            }
        }

        return String.join(" ", result);
    }

    // 역직렬화
    public TreeNode deserialize(String data) {
        // 예외 처리
        if (data.equals("# #")) {
            return null;
        }

        String[] nodes = data.split(" ");

        TreeNode root = new TreeNode(Integer.parseInt(nodes[1]));

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        int index = 2;

        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();

            // 왼쪽 자식
            if (!nodes[index].equals("#")) {
                node.left = new TreeNode(Integer.parseInt(nodes[index]));
                queue.offer(node.left);
            }
            index++;

            // 오른쪽 자식
            if (!nodes[index].equals("#")) {
                node.right = new TreeNode(Integer.parseInt(nodes[index]));
                queue.offer(node.right);
            }
            index++;
        }

        return root;
    }
}

File: 0297-serialize-and-deserialize-binary-tree/0297-serialize-and-deserialize-binary-tree.ts
// 직렬화
function serialize(root: TreeNode | null): string {
    if (root === null) {
        return "# #";
    }

    const queue: (TreeNode | null)[] = [root];
    const result: string[] = ["#"];

    let index = 0;

    // BFS
    while (index < queue.length) {
        const node = queue[index++];

        if (node !== null) {
            queue.push(node.left);
            queue.push(node.right);

            result.push(String(node.val));
        } else {
            result.push("#");
        }
    }

    return result.join(" ");
}


// 역직렬화
function deserialize(data: string): TreeNode | null {
    if (data === "# #") {
        return null;
    }

    const nodes: string[] = data.split(" ");

    const root = new TreeNode(Number(nodes[1]));

    const queue: TreeNode[] = [root];

    let queueIndex = 0;
    let index = 2;

    // BFS
    while (queueIndex < queue.length) {
        const node = queue[queueIndex++];

        // 왼쪽 자식
        if (nodes[index] !== "#") {
            node.left = new TreeNode(Number(nodes[index]));
            queue.push(node.left);
        }

        index++;

        // 오른쪽 자식
        if (nodes[index] !== "#") {
            node.right = new TreeNode(Number(nodes[index]));
            queue.push(node.right);
        }

        index++;
    }

    return root;
}